File: app/src/main/java/com/vkpapps/soundbooster/connection/ClientHelper.java
package com.vkpapps.soundbooster.connection;

import android.os.Bundle;
import android.util.Log;

import androidx.annotation.NonNull;

import com.vkpapps.soundbooster.interfaces.OnClientConnectionStateListener;
import com.vkpapps.soundbooster.interfaces.OnControlRequestListener;
import com.vkpapps.soundbooster.interfaces.OnObjectReceiveListener;
import com.vkpapps.soundbooster.model.User;
import com.vkpapps.soundbooster.model.control.ControlFile;
import com.vkpapps.soundbooster.model.control.ControlPlayer;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.net.Socket;

public class ClientHelper extends Thread {
    public User user;
    private ObjectOutputStream outputStream;
    private Socket socket;
    private OnControlRequestListener onControlRequestListener;
    private OnClientConnectionStateListener onClientConnectionStateListener;
    private OnObjectReceiveListener onObjectReceiveListener;

    public ClientHelper(Socket socket, OnControlRequestListener onControlRequestListener, User user, OnClientConnectionStateListener onClientConnectionStateListener) {
        this.socket = socket;
        this.user = user;
        this.onControlRequestListener = onControlRequestListener;
        this.onClientConnectionStateListener = onClientConnectionStateListener;
    }

    @Override
    public void run() {
        Bundle bundle = new Bundle();
        bundle.putString("ID", user.getUserId());
        String TAG = "ClientHelper";
        try {
            Log.d(TAG, "run: ==================================== connecting...to istream ");
            outputStream = new ObjectOutputStream(socket.getOutputStream());
            // send identity to connected device
            outputStream.writeObject(user);
            outputStream.flush();

            ObjectInputStream inputStream = new ObjectInputStream(socket.getInputStream());
            Object object = inputStream.readObject();
            Log.d(TAG, "run: ==================================== connect " + user);
            if (object instanceof User) {

                user = (User) object;

                //notify user added
                if (onClientConnectionStateListener != null) {
                    onClientConnectionStateListener.onClientConnected(this);
                }

                while (socket.isConnected()) {
                    object = inputStream.readObject();
                    if (object instanceof ControlPlayer) {
                        ControlPlayer controlPlayer = (ControlPlayer) object;
                        handleControl(controlPlayer);
                    } else if (object instanceof ControlFile) {
                        ControlFile control = (ControlFile) object;
                        handleFileControl(control);
                    } else if (object instanceof User) {
                        User u = (User) object;
                        if (u.getUserId().equals(user.getUserId())) {
                            user.setAccess(u.isAccess());
                            user.setName(u.getName());
                        }
                    } else {
                        System.err.println("invalid " + object);
                    }
                }

            } else {
                return;
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        // notify client leaved or disconnected
        Log.d(TAG, "run: ==================================== disconnect ");
        if (onClientConnectionStateListener != null) {
            onClientConnectionStateListener.onClientDisconnected(this);
        }
    }

    public void write(@NonNull Object command) {
        new Thread(() -> {
            try {
                outputStream.writeObject(command);
                outputStream.flush();
            } catch (IOException e) {
                e.printStackTrace();
            }
        }).start();
    }

    public void writeIfClient(@NonNull Object command, String clientId) {
        new Thread(() -> {
            if (user.getUserId().equals(clientId)) {
                try {
                    outputStream.writeObject(command);
                    outputStream.flush();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }).start();
    }

    private void handleControl(ControlPlayer control) {

        if (onObjectReceiveListener != null) {
            onObjectReceiveListener.onObjectReceive(control);
        }
        Log.d("CONTROLS", "handleMessage: =================================== req " + control.getAction());
        onControlRequestListener.onMusicPlayerControl(control);
    }

    private void handleFileControl(ControlFile control) {
        try {
            Log.d("CONTROLS", "handleMessage: =================================== req " + control.getAction());
            switch (control.getAction()) {
                case ControlFile.DOWNLOAD_REQUEST:
                    onControlRequestListener.onDownloadRequest(control.getData(), control.getId());
                    break;
                case ControlFile.UPLOAD_REQUEST:
                    onControlRequestListener.onUploadRequest(control.getData(), control.getId());
                    break;
                case ControlFile.DOWNLOAD_REQUEST_CONFIRM:
                    onControlRequestListener.onDownloadRequestAccepted(control.getData(), control.getId());
                    break;
                case ControlFile.UPLOAD_REQUEST_CONFIRM:
                    onControlRequestListener.onUploadRequestAccepted(control.getData(), control.getId());
                    break;
                default:
                    Log.d("CONTROLS", "handleMessage: =================================== invalid req ");
            }
        } catch (NullPointerException | ArrayIndexOutOfBoundsException e) {
            e.printStackTrace();
        }
    }

    public void setOnObjectReceiveListener(OnObjectReceiveListener onObjectReceiveListener) {
        this.onObjectReceiveListener = onObjectReceiveListener;
    }
}

File: app/src/main/java/com/vkpapps/soundbooster/connection/ClientHelper.kt
package com.vkpapps.soundbooster.connection

import android.os.Bundle
import com.vkpapps.soundbooster.analitics.Logger
import com.vkpapps.soundbooster.interfaces.OnClientConnectionStateListener
import com.vkpapps.soundbooster.interfaces.OnControlRequestListener
import com.vkpapps.soundbooster.interfaces.OnObjectReceiveListener
import com.vkpapps.soundbooster.model.User
import com.vkpapps.soundbooster.model.control.ControlFile
import com.vkpapps.soundbooster.model.control.ControlPlayer
import java.io.IOException
import java.io.ObjectInputStream
import java.io.ObjectOutputStream
import java.net.Socket

/***
 * @author VIJAY PATIDAR
 */
class ClientHelper(private val socket: Socket, private val onControlRequestListener: OnControlRequestListener, var user: User, private val onClientConnectionStateListener: OnClientConnectionStateListener?) : Thread() {
    private var outputStream: ObjectOutputStream? = null
    private var onObjectReceiveListener: OnObjectReceiveListener? = null
    override fun run() {
        val bundle = Bundle()
        bundle.putString("ID", user.userId)
        try {
            outputStream = ObjectOutputStream(socket.getOutputStream())
            // send identity to connected device
            outputStream!!.writeObject(user)
            outputStream!!.flush()
            val inputStream = ObjectInputStream(socket.getInputStream())
            var obj = inputStream.readObject()
            if (obj is User) {
                user = obj

                //notify user added
                onClientConnectionStateListener?.onClientConnected(this)
                var retry = 0
                while (!socket.isClosed) {
                    try {
                        obj = inputStream.readObject()
                        if (obj is ControlPlayer) {
                            onControlRequestListener.onMusicPlayerControl(obj)
                            onObjectReceiveListener?.onObjectReceive(obj)
                        } else if (obj is ControlFile) {
                            handleFileControl(obj)
                        } else if (obj is User) {
                            // update user information
                            if (obj.userId == user.userId) {
                                user.name = obj.name
                            }
                        } else {
                            Logger.e("invalid object received $obj")
                        }
                    } catch (e: Exception) {
                        retry++
                        if (retry == 10) break
                        e.printStackTrace()
                    }
                }
            } else {
                return
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        // notify client leaved or disconnected
        onClientConnectionStateListener?.onClientDisconnected(this)
    }

    fun write(command: Any) {
        Thread(Runnable {
            try {
                outputStream?.writeObject(command)
                outputStream?.flush()
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }).start()
    }

    private fun handleFileControl(control: ControlFile) {
        try {
            when (control.action) {
                ControlFile.DOWNLOAD_REQUEST -> onControlRequestListener.onDownloadRequest(control.data, control.id, control.type)
                ControlFile.UPLOAD_REQUEST -> onControlRequestListener.onUploadRequest(control.data, control.id, control.type)
                ControlFile.DOWNLOAD_REQUEST_CONFIRM -> onControlRequestListener.onDownloadRequestAccepted(control.data, control.id, control.type)
                ControlFile.UPLOAD_REQUEST_CONFIRM -> onControlRequestListener.onUploadRequestAccepted(control.data, control.id, control.type)
                else -> Logger.d("handleFileControl: invalid req " + control.action)
            }
        } catch (e: NullPointerException) {
            e.printStackTrace()
        } catch (e: ArrayIndexOutOfBoundsException) {
            e.printStackTrace()
        }
    }

    fun setOnObjectReceiveListener(onObjectReceiveListener: OnObjectReceiveListener) {
        this.onObjectReceiveListener = onObjectReceiveListener
    }

    fun shutDown() {
        try {
            socket.close()
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

}